rand_distr/src/poisson.rs

```diff
@@ -104,77 +127,99 @@ where
         if !(lambda > F::zero()) {
             return Err(Error::ShapeTooSmall);
         }
-        let log_lambda = lambda.ln();
-        Ok(Poisson {
-            lambda,
-            exp_lambda: (-lambda).exp(),
-            log_lambda,
-            sqrt_2lambda: (F::from(2.0).unwrap() * lambda).sqrt(),
-            magic_val: lambda * log_lambda - crate::utils::log_gamma(F::one() + lambda),
-        })
+
+        // Use the Knuth method only for low expected values
+        let method = if lambda < F::from(12.0).unwrap() {
+            Method::Knuth(KnuthMethod::new(lambda))
+        } else {
+            let log_lambda = lambda.ln();
+            let sqrt_2lambda = (F::from(2.0).unwrap() * lambda).sqrt();
+            let magic_val = lambda * log_lambda - crate::utils::log_gamma(F::one() + lambda);
+            Method::Rejection(RejectionMethod {
+                log_lambda,
+                sqrt_2lambda,
+                magic_val,
+            })
+        };
+
+        Ok(Poisson { lambda, method })
     }
 }
 
-impl<F> Distribution<F> for Poisson<F>
+impl<F> Distribution<F> for KnuthMethod<F>
 where
     F: Float + FloatConst,
     Standard: Distribution<F>,
 {
-    #[inline]
     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> F {
-        // using the algorithm from Numerical Recipes in C
-
-        // for low expected values use the Knuth method
-        if self.lambda < F::from(12.0).unwrap() {
-            let mut result = F::one();
-            let mut p = rng.random::<F>();
-            while p > self.exp_lambda {
-                p = p * rng.random::<F>();
-                result = result + F::one();
-            }
-            result - F::one()
+        let mut result = F::one();
+        let mut p = rng.random::<F>();
+        while p > self.exp_lambda {
+            p = p * rng.random::<F>();
+            result = result + F::one();
         }
-        // high expected values - rejection method
-        else {
-            // we use the Cauchy distribution as the comparison distribution
-            // f(x) ~ 1/(1+x^2)
-            let cauchy = Cauchy::new(F::zero(), F::one()).unwrap();
-            let mut result;
+        result - F::one()
+    }
+}
+impl<F> RejectionMethod<F>
+where
+    F: Float + FloatConst,
+    Standard: Distribution<F>,
+{
+    fn sample<R: Rng + ?Sized>(&self, lambda: F, rng: &mut R) -> F {
```

Comment:
I don't really like that `RejectionMethod` uses a non-standard `sample` signature, even if its not publicly visible. I think anything that implements `sample` should own all the attributes it needs to be sampled.

Reply:
This is a reasonable point, though since it's internal to the module I don't think it important.

Reply:
Yes fair enough, it just makes me feel a little unclean is all, considering there is only one value for the parameter which ever leads to sound behavior.

Reply:
I agree, this would also allow `RejectionMethod` to implement `Distribution` which could be useful somewhere else.
The only advantage of the current code is that a `get_lambda` function would be easier.